File: packages/python/anip-server/src/anip_server/retention_enforcer.py
"""Retention enforcer -- background cleanup of expired audit entries (v0.8)."""

from __future__ import annotations

import asyncio
import time
from collections.abc import Callable
from datetime import datetime, timezone
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from .storage import StorageBackend


class RetentionEnforcer:
    """Background sweep that deletes expired audit entries."""

    def __init__(
        self,
        storage: StorageBackend,
        *,
        interval_seconds: int = 60,
        skip_audit_retention: bool = False,
        on_sweep: Callable[[int, float], None] | None = None,
        on_error: Callable[[str], None] | None = None,
    ) -> None:
        self._storage = storage
        self._interval = interval_seconds
        self._skip_audit_retention = skip_audit_retention
        self._on_sweep = on_sweep
        self._on_error = on_error
        self._task: asyncio.Task[None] | None = None
        self._last_run_at: str | None = None
        self._last_deleted_count: int = 0
        self._last_error: str | None = None

    @property
    def last_run_at(self) -> str | None:
        return self._last_run_at

    @property
    def last_deleted_count(self) -> int:
        return self._last_deleted_count

    @property
    def last_error(self) -> str | None:
        return self._last_error

    @property
    def is_running(self) -> bool:
        return self._task is not None and not self._task.done()

    async def sweep(self) -> int:
        """Run one cleanup sweep. Returns number of deleted entries."""
        if self._skip_audit_retention:
            return 0  # Cluster mode: audit retention disabled
        now = datetime.now(timezone.utc).isoformat()
        return await self._storage.delete_expired_audit_entries(now)

    def start(self) -> None:
        """Start background cleanup as an asyncio task.

        Must be called from within a running event loop.
        """
        try:
            loop = asyncio.get_running_loop()
        except RuntimeError:
            raise RuntimeError(
                "RetentionEnforcer.start() requires a running event loop. "
                "Call from an async context (e.g., ASGI startup hook)."
            )
        self._task = loop.create_task(self._run())

    def stop(self) -> None:
        """Cancel the background cleanup task."""
        if self._task is not None:
            self._task.cancel()
            self._task = None

    async def _run(self) -> None:
        while True:
            await asyncio.sleep(self._interval)
            start = time.monotonic()
            try:
                count = await self.sweep()
                duration_ms = (time.monotonic() - start) * 1000
                self._last_run_at = datetime.now(timezone.utc).isoformat()
                self._last_deleted_count = count
                self._last_error = None
                if self._on_sweep:
                    try:
                        self._on_sweep(count, duration_ms)
                    except Exception:
                        pass
            except asyncio.CancelledError:
                raise
            except Exception as e:
                self._last_error = str(e)
                if self._on_error:
                    try:
                        self._on_error(str(e))
                    except Exception:
                        pass

File: packages/typescript/server/src/retention-enforcer.ts
/**
 * Retention enforcer -- background cleanup of expired audit entries (v0.8).
 */
import type { StorageBackend } from "./storage.js";

export interface RetentionEnforcerOpts {
  skipAuditRetention?: boolean;
  onSweep?: (deletedCount: number, durationMs: number) => void;
  onError?: (error: string) => void;
}

export class RetentionEnforcer {
  private _storage: StorageBackend;
  private _interval: number;
  private _timer: ReturnType<typeof setInterval> | null = null;
  private _skipAuditRetention: boolean;
  private _onSweep?: (deletedCount: number, durationMs: number) => void;
  private _onError?: (error: string) => void;
  private _lastRunAt: string | null = null;
  private _lastDeletedCount: number = 0;
  private _lastError: string | null = null;

  constructor(storage: StorageBackend, intervalSeconds: number = 60, opts?: RetentionEnforcerOpts) {
    this._storage = storage;
    this._interval = intervalSeconds;
    this._skipAuditRetention = opts?.skipAuditRetention ?? false;
    this._onSweep = opts?.onSweep;
    this._onError = opts?.onError;
  }

  getLastRunAt(): string | null { return this._lastRunAt; }
  getLastDeletedCount(): number { return this._lastDeletedCount; }
  getLastError(): string | null { return this._lastError; }

  isRunning(): boolean {
    return this._timer !== null;
  }

  async sweep(): Promise<number> {
    if (this._skipAuditRetention) return 0;
    const now = new Date().toISOString();
    return this._storage.deleteExpiredAuditEntries(now);
  }

  start(): void {
    if (this._timer) return;
    this._timer = setInterval(async () => {
      const start = performance.now();
      try {
        const count = await this.sweep();
        const durationMs = Math.round(performance.now() - start);
        this._lastRunAt = new Date().toISOString();
        this._lastDeletedCount = count;
        this._lastError = null;
        this._onSweep?.(count, durationMs);
      } catch (e: unknown) {
        this._lastError = String(e);
        this._onError?.(String(e));
      }
    }, this._interval * 1000);
    // Allow the timer to not prevent process exit (matches CheckpointScheduler)
    if (this._timer && typeof this._timer === "object" && "unref" in this._timer) {
      (this._timer as NodeJS.Timeout).unref();
    }
  }

  stop(): void {
    if (this._timer) {
      clearInterval(this._timer);
      this._timer = null;
    }
  }
}
